components/renku_data_services/base_api/auth.py

```diff
@@ -153,3 +244,30 @@ async def decorated_function(*args: _P.args, **kwargs: _P.kwargs) -> _T:
         return response
 
     return decorated_function
+
+
+def internal_gitlab_authenticate(
```

Comment:
suggestion: I think we could rewrite the decorator to accept chained authenticators, and the authenticators can wrap the user objects with dedicated objects (Like GitUser(APIUser()) or similar, if their `authenticate` method accepted an optional `user` object from previous calls) to be more flexible. That would allow adding additional authenticators more easily without an explosion of custom decorators, but I'm a bit torn on if it actually makes the code simpler/easier to understand.

Reply:
Are you thinking of something like:
```
@authenticate([APIUser, GitUser])
```
Passing an ordered list of authenticator my be more readable

Reply:
well I guess it'd be `@authenticate([self.authenticator, self.gitlab_authenticator])`, the `GitUser(ApiUser())` was more of an example of what the return value of that might be. Either using a decorator pattern (classes with the same base api wrapping each other) or just composition in general? Could also just be adding to some context object on APIUser. as an end-result, it'd just be nice to have one object that can carry all the user authentication related information (especially since typing decorators is such a pain and having so do stuff like "the api user is always in position 2 and the gitlab user is always in position 3" is pretty annoying and conflict prone)

Reply:
> well I guess it'd be @authenticate([self.authenticator, self.gitlab_authenticator])

The problem with this getting mypy and the typing right on such a decorator. And this is possible but it will probably mean that you have to inject the APIUser instances in the decorated functions as a list. And then you lose information about how many there are and what they are.

That is why I went for this:

```
def authenticate_2(
    authenticator1: Authenticator,
    authenticator2: Authenticator,
) -> Callable[
    [Callable[Concatenate[Request, AnyAPIUser, AnyAPIUser, _P], Coroutine[Any, Any, _T]]],
    Callable[Concatenate[Request, _P], Coroutine[Any, Any, _T]],
]:
    """Decorator for a Sanic handler that adds the APIUser when another authentication has already been done."""

    def decorator(
        f: Callable[Concatenate[Request, AnyAPIUser, AnyAPIUser, _P], Coroutine[Any, Any, _T]],
    ) -> Callable[Concatenate[Request, _P], Coroutine[Any, Any, _T]]:
        @wraps(f)
        async def decorated_function(request: Request, *args: _P.args, **kwargs: _P.kwargs) -> _T:
            token1 = request.headers.get(authenticator1.token_field)
            token2 = request.headers.get(authenticator2.token_field)
            user1: AnyAPIUser
            user2: AnyAPIUser
            [user1, user2] = await asyncio.gather(
                authenticator1.authenticate(token1 or "", request),
                authenticator2.authenticate(token2 or "", request),
            )
            response = await f(request, user1, user2, *args, **kwargs)
            return response

        return decorated_function

    return decorator
```

This way we retain all the useful info about the stuff being injected in the function. And we insure that if we want 2 authenticator then we get 2 APIUsers injected.

I dont think we will need that many but if really needed we can make `_3` and similar variants. But I think the most we will need is 2.

Reply:
I already spent 3-4 hours trying to get the typing right on the above example which is simpler IMO (when it comes to mypy) than what you propose. I really dont want to spend more time wrestling with mypy.